planner/cascades/transformation_rules.go

```diff
@@ -599,6 +600,40 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
 
+// PushSelDownUnionAll pushes selection through union all.
+type PushSelDownUnionAll struct {
+	baseRule
+}
+
+// NewRulePushSelDownUnionAll creates a new Transformation PushSelDownUnionAll.
+// The pattern of this rule is `Selection -> UnionAll`.
+func NewRulePushSelDownUnionAll() Transformation {
+	rule := &PushSelDownUnionAll{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandSelection,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandUnionAll, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// It will transform `Selection->UnionAll->x` to `UnionAll->Selection->x`.
+func (r *PushSelDownUnionAll) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
+	unionAll := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalUnionAll)
+	childGroup := old.Children[0].GetExpr().Children[0]
+
+	newSelExpr := memo.NewGroupExpr(sel)
+	newSelExpr.Children = append(newSelExpr.Children, childGroup)
+	newSelGroup := memo.NewGroupWithSchema(newSelExpr, childGroup.Prop.Schema)
+
+	newUnionAllExpr := memo.NewGroupExpr(unionAll)
+	newUnionAllExpr.Children = append(newUnionAllExpr.Children, newSelGroup)
+
+	return []*memo.GroupExpr{newUnionAllExpr}, true, false, nil
+}
```

Comment:
UnionAll has more than one child. You should push the selection down to all of it child. For example:
`Selection -> UnionAll -> (Child0, Child1, Child2)` to `UnionAll -> (Selection -> Child0, Selection -> Child1, Selection -> Child2)`

Reply:
`OnTransform` should look like:
```
for i := range oldChildGroups {
    create newSelExpr
    newSelExpr.Child = oldChildGroups[i]
    create newChildGroup from newSelExpr
    newChildGroups[i] = newChildGroup
}
create newUnionAllExpr
set newUnionAllExpr's children as newChildGroups
return newUnionAllExpr
```

Reply:
Thank you, how about this:

    sel := old.GetExpr().ExprNode.(*plannercore.LogicalSelection)
	unionAll := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalUnionAll)
	childGroups := old.Children[0].GetExpr().Children

	newUnionAllExpr := memo.NewGroupExpr(unionAll)
	for _, group := range childGroups {
		newSelExpr := memo.NewGroupExpr(sel)
		newSelExpr.Children = append(newSelExpr.Children, group)
		newSelGroup := memo.NewGroupWithSchema(newSelExpr, group.Prop.Schema)

		newUnionAllExpr.Children = append(newUnionAllExpr.Children, newSelGroup)
	}
	return []*memo.GroupExpr{newUnionAllExpr}, true, false, nil

Reply:
LGTM